uamqp/client.py

```diff
@@ -639,14 +638,30 @@ def receive_message_batch(self, max_batch_size=None, on_message_received=None, t
         else:
             return batch
 
-        while receiving and not expired and len(batch) < max_batch_size:
-            receiving = self.do_work(batch=max_batch_size)
-            while len(batch) < max_batch_size:
-                try:
-                    batch.append(self._received_messages.get_nowait())
-                    self._received_messages.task_done()
-                except queue.Empty:
-                    break
+        to_receive_size = max_batch_size - len(batch)
+        before_queue_size = self._received_messages.qsize()
+
+        while receiving and to_receive_size > 0:
             if timeout and time.time() > timeout:
-                expired = True
+                break
+
+            receiving = self.do_work(batch=to_receive_size)
+            cur_queue_size = self._received_messages.qsize()
+            # after do_work, check how many new messages have been received since previous iteration
+            received = cur_queue_size - before_queue_size
+            if to_receive_size < max_batch_size and received == 0:
```

Comment:
I'm not sure I follow what we're checking with this condition: `if to_receive_size < max_batch_size`

Reply:
Oh actually I think I get - so if the batch was initially empty, `to_receive_size` == `max_batch_size`.
So we just want to avoid breaking on an empty batch. Yup I think this makes sense :)

Reply:
you got it!